executor/show_test.go

```diff
@@ -312,6 +312,36 @@ func (s *testSuite2) TestShowCreateTable(c *C) {
 
 	tk.MustExec("drop view v1")
 	tk.MustExec("drop table t1")
+
+	// For issue #9211
+	tk.MustExec("create table t(c int, b int as (c + 1))ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;")
+	tk.MustQuery("show create table `t`").Check(testutil.RowsWithSep("|",
+		""+
+			"t CREATE TABLE `t` (\n"+
+			"  `c` int(11) DEFAULT NULL,\n"+
+			"  `b` int(11) GENERATED ALWAYS AS (`c` + 1) VIRTUAL\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin",
+	))
+
+	tk.MustExec("drop table t")
+	tk.MustExec("create table t(c int, b int as (c + 1) not null)ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;")
+	tk.MustQuery("show create table `t`").Check(testutil.RowsWithSep("|",
+		""+
+			"t CREATE TABLE `t` (\n"+
+			"  `c` int(11) DEFAULT NULL,\n"+
+			"  `b` int(11) GENERATED ALWAYS AS (`c` + 1) VIRTUAL NOT NULL\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin",
+	))
+	tk.MustExec("drop table t")
+	tk.MustExec("create table t ( a char(10) charset utf8 collate utf8_bin, b char(10) as (rtrim(a)));")
+	tk.MustQuery("show create table `t`").Check(testutil.RowsWithSep("|",
+		""+
+			"t CREATE TABLE `t` (\n"+
+			"  `a` char(10) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL,\n"+
+			"  `b` char(10) GENERATED ALWAYS AS (rtrim(`a`)) VIRTUAL\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin",
+	))
+	tk.MustExec("drop table t")
 }
```

Comment:
I was wondering if you could add test like this `create table test ( a char(10) charset gb2312 , b char(10) charset utf8 as (rtrim(a)) );`?
BTW, The following is the result of MYSQL:
```sql
mysql> create table test ( a char(10) charset gb2312 , b char(10) charset utf8 as (rtrim(a)) );
Query OK, 0 rows affected (0.02 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                              |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `a` char(10) CHARACTER SET gb2312 DEFAULT NULL,
  `b` char(10) CHARACTER SET utf8 GENERATED ALWAYS AS (rtrim(`a`)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 5.7.25-0ubuntu0.18.04.2 |
+-------------------------+
1 row in set (0.00 sec)
```

Reply:
@xiekeyi98 I seems make a mistake here. I made several tests and found that whether a column is generated do not actually determine whether `CHARSET` && `COLLATE` are shown.  In fact, it's determined by whether `CHARSET` && `COLLATE` are specified explicitly.  
I'd prefer left it to another issue and make this pr clearer.

BTW, `TiDB` seems don't support `gb2312` for now.
The following are my tests:
```sql
mysql> select version();
+------------+
| version()  |
+------------+
| 5.7.17-log |
+------------+
1 row in set (0.01 sec)

mysql> create table test ( a char(10) charset utf8 , b char(10) charset utf8 as (rtrim(a)) );
Query OK, 0 rows affected (0.03 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------+| Table | Create Table                                                                                                                        |+-------+-------------------------------------------------------------------------------------------------------------------------------------+| test  | CREATE TABLE `test` (
  `a` char(10) CHARACTER SET utf8 DEFAULT NULL,
  `b` char(10) CHARACTER SET utf8 GENERATED ALWAYS AS (rtrim(`a`)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+1 row in set (0.00 sec)

mysql> drop table test;
Query OK, 0 rows affected (0.02 sec)

mysql> create table test ( a char(10) charset utf8 , b char(10) as (rtrim(a)) );
Query OK, 0 rows affected (0.03 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                        |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `a` char(10) CHARACTER SET utf8 DEFAULT NULL,
  `b` char(10) GENERATED ALWAYS AS (rtrim(`a`)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
mysql> drop table test;
Query OK, 0 rows affected (0.02 sec)

mysql> create table test ( a char(10), b char(10) as (rtrim(a)) );
Query OK, 0 rows affected (0.03 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                        |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `a` char(10) DEFAULT NULL,
  `b` char(10) GENERATED ALWAYS AS (rtrim(`a`)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> desc test;
+-------+----------+------+-----+---------+-------------------+
| Field | Type     | Null | Key | Default | Extra             |
+-------+----------+------+-----+---------+-------------------+
| a     | char(10) | YES  |     | NULL    |                   |
| b     | char(10) | YES  |     | NULL    | VIRTUAL GENERATED |
+-------+----------+------+-----+---------+-------------------+
2 rows in set (0.01 sec)

```

Reply:
@xiekeyi98 I've opened another issue #9232  for the demonstration of the demonstration of `CHARSET` && `COLLATE`.  It's not related to generated columns, indeed. I'll look into this later.

Reply:
Well done. Thanks for you contribution.

Reply:
seems you've working on this in #9113  :P